Onelogin (#1)
taksan/authenticator-saml-onelogin: README.md

# SAML 2.0 Authenticator 

This module has been developed to provide a working SAML 2.0 Authenticator and is mostly based on
[https://github.com/xwiki-contrib/authenticator-saml](https://github.com/xwiki-contrib/authenticator-saml), 

> but replaces OpenSAML (which has reached its End of Life) by OneLogin [https://github.com/onelogin/java-saml](https://github.com/onelogin/java-saml) 

Reviewer: There's no reason for a markdown pretty link where the "pretty" part is the link itself. Either replace the text with a pretty version or remove the link altogether.

for the SAML authentication.

## TL;DR

This module has been tested with the follow Identity Providers. Follow the instructions below
for assistance for proper configuration.

[Google Workspace](google-workspace-setup.md)

[KeyCloak](keycloak-setup.md)

> If you set up this plugin with other IDPs, please feel free to contribute opening PR's

Reviewer: This is not a possessive, no apostrophe:
```suggestion
If you set up this plugin with other IDPs, please feel free to contribute opening PRs.
```

It's complicated, read [this](https://www.thepunctuationguide.com/apostrophe.html) if you need a reference.

with your instructions.

## Dev Setup Quick Start

If you want to test the plugin by yourself, go to our [Dev Quick Start](dev-quick-start.md) and follow
the instructions to set up a complete local environment. 

# Reference of plugin configurations on xwiki.cfg

Once you install the plugin in your XWiki instance, to make it work properly you will need to set up 
a couple of properties.

## Required properties

> xwiki.authentication.authclass=com.xwiki.authentication.saml.XWikiSAML20Authenticator

Reviewer: Should be pre-formatted with backticks, either single backticks for inline \`like this\` (`like this`) or triple backticks for multiline:
\```  
like this
\```
```
like this
```

Also... What is this?

### xwiki.authentication.saml2.idp.single_sign_on_service.url

> IDP stands for Identity Provider, and is the service that provides the authentication, for example, Google Workspace, KeyCloak, and others like it.

Reviewer: This information is useless for anyone reading up to this point, everyone that got to this line will already know what IDP stands for.

> The single_sign_on_service is used to tell what is the IDP URL where authentication happens. Usually, you can get this information from metadata.xml. Choose the "POST" URL.

Reviewer: `The single_sign_on_service is used` -> `This option is used`

You are already "under" the title, no need to repeat the thing you are talking about.

### xwiki.authentication.saml2.idp.entityid

The IDP entity id is a unique string that identifies the IDP and is used by the

> authenticator to validate the IDP. You usually get this information from a metadata.xml file.

Reviewer: ```suggestion
The IDP entity ID is a unique string that identifies the IDP and is used by the
authenticator to validate the IDP. You usually get this information from a `metadata.xml` file.
```

### xwiki.authentication.saml2.sp.entityid

SP stands for "Service Provider", in this case the XWiki instance. The SP entity id
is a unique string (it can be any arbitrary value chosen by you) that identifies

> your XWiki instance. Some people recommend using the application URL.

Reviewer: ```suggestion
SP stands for "Service Provider", in this case the XWiki instance. The SP entity ID
is a unique string (it can be any arbitrary value chosen by you) that identifies
your XWiki instance. Some people recommend using the application URL.
```

Usually, you will configure this ID somewhere in the IDP, so it can tell whether authorization 
requests are coming from a valid application and to choose the correct configuration.

### xwiki.authentication.saml2.idp.x509cert

This is the IDP certificate. This is used to validate the identity of the IDP provider
to ensure it is legitimate.

### xwiki.authentication.saml2.sp.assertion_consumer_service.url

This is XWiki's URL that accepts the post to process the authentication coming from IDP. It will always be like this:

>     https://<you wiki domain>/bin/loginsubmit/XWiki/XWikiLogin

Reviewer: This isn't formatted, so it was broken on parsing. Use a multiline format block:
````suggestion
```
    https://<your wiki domain>/bin/loginsubmit/XWiki/XWikiLogin
```
````
Also, `you` -> `your`

> 

Reviewer: Using the options as titles is kinda "meh"? It looks less like code and it's definitely not a title.

Suggestions: 
1. Transform it into a list with explanations, like so:
  * `xwiki.authentication.saml2.idp.single_sign_on_service.url`:
    This option does a thing and something else and etc etc some other bla bla bla text.  
    Can even have more lines.
  * `xwiki.authentication.saml2.idp.x509cert`:
    This other option is very nice, it makes things not break.
  
---
2. A table:

| Option | Description |
|---|---|
| `xwiki.authentication.saml2.idp.single_sign_on_service.url` | This option does a thing and something else and etc etc some other bla bla bla text. Can't have line splits on this one, it will split whenever it pleases. |
| `xwiki.authentication.saml2.idp.x509cert` | This other option is very nice, it makes things not break. |

---

Markdown for these two:
```markdown
1. Transform it into a list with explanations, like so:
  * `xwiki.authentication.saml2.idp.single_sign_on_service.url`:
    This option does a thing and something else and etc etc some other bla bla bla text.  
    Can even have more lines.
  * `xwiki.authentication.saml2.idp.x509cert`:
    This other option is very nice, it makes things not break.
  
---
2. A table:

| Option | Description |
|---|---|
| `xwiki.authentication.saml2.idp.single_sign_on_service.url` | This option does a thing and something else and etc etc some other bla bla bla text. Can't have line splits on this one, it will split whenever it pleases. |
| `xwiki.authentication.saml2.idp.x509cert` | This other option is very nice, it makes things not break. |

---

```

## Properties with default values

The following properties already have default values that will work for most cases and don't need
to be changed, but can be customized in case you have very specific needs.

### Mapping of XWiki user fields and identity provider
    xwiki.authentication.saml2.fields_mapping=email=email,first_name=firstName,last_name=lastName

Map fields provided by the IDP with XWiki fields. The format is like this:

>     <xwiki field name>=<idp field name>

Reviewer: This also doesn't show on the parsed README, use a multiline format block

> You can map multiple fields separating each mapping by comma.

Reviewer: Place the example under the explanation, not the other way around. Also, avoid using 4 spaces to format code, it is quite finicky and prone to breaking. Just use \```

The default value means:

+ IDP's email field will be mapped to Xwiki's email
+ IDP's firstName field will be mapped to Xwiki's first_name

> + IDP's lastName field will be mapped to Xwiki's last_name

Reviewer: Inline code block the fields:
* IDP's `email` field will be mapped to Xwiki's `email`

Same for the rest

As a rule of thumb, anything that is code, configuration, whatever that is not actual text, format with code blocks. It catches the eye, makes it easier to distinguish and your documentation more easily interpreted. Also facilitates copy and paste.

The above fields are the minimum required fields.

### Default group for new users

>     xwiki.authentication.saml2.defaultGroupForNewUsers=XWiki.SamlGroup

Reviewer: Same suggestion as above, avoid 4 spaces for formatting, it's prone to breaking without anyone noticing. Same for all below.

You can set up the default group for users coming from the IDP. 

### The name of the attribute used to cache the authentication result in the current session
    xwiki.authentication.saml2.auth_field=saml_user

You usually don't need to bother with this configuration. It is used to save a custom field
in the XWiki's users, mapping the IDP's user identity into it. Once you start with a value,
we don't recommend changing it. 

### List of fields to use for generating an XWiki username
    xwiki.authentication.saml2.xwiki_user_rule=first_name,last_name

This property controls how the plugin will generate the username in XWiki, based on the user's
field values.

### Capitalize each field value when generating the username
    xwiki.authentication.saml2.xwiki_user_rule_capitalize=true

Controls whether username generation should capitalize each of the  xwiki_user_rule's attributes composition

### NameIDFormat format
    xwiki.authentication.saml2.sp.nameidformat=urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress

This field determines which IDP's field will be used to identify the user in XWiki. When a user authenticates, 

> this value will be stored in the field specified by ```xwiki.authentication.saml2.auth_field``` property

Reviewer: ```suggestion
this value will be stored in the field specified by `xwiki.authentication.saml2.auth_field` property
```

> 

Reviewer: Trailing blank lines. Leave just the final newline.

# SAML 2.0 Authenticator 

This module has been developed to provide a working SAML 2.0 Authenticator and is mostly based on

> https://github.com/xwiki-contrib/authenticator-saml, 

Reviewer: ```suggestion
[xwiki-contrib/authenticator-saml](https://github.com/xwiki-contrib/authenticator-saml), 
```

> but replaces OpenSAML (which has reached its End of Life) by OneLogin https://github.com/onelogin/java-saml 

Reviewer: ```suggestion
but replaces OpenSAML (which has reached its End of Life) by [OneLogin](https://github.com/onelogin/java-saml) 
```

for the SAML authentication.

## TL;DR

This module has been tested with the follow Identity Providers. Follow the instructions below
for assistance for proper configuration.

[Google Workspace](google-workspace-setup.md)

> [KeyCloak](keycloak-setup.md)

Reviewer: ```suggestion
* [Google Workspace](google-workspace-setup.md)
* [KeyCloak](keycloak-setup.md)
```

If you set up this plugin with other IDPs, please feel free to contribute opening PRs.

> with your instructions.

Reviewer: ```suggestion
If you set up this plugin with other IDPs, please feel free to contribute by opening PRs
with your instructions.
```

## Dev Setup Quick Start

If you want to test the plugin by yourself, go to our [Dev Quick Start](quick-start/README.md) and follow

> the instructions to set up a complete local environment. 

Reviewer: ```suggestion
If you want to modify the source code or help develop the plugin, go to our [Dev Quick Start](quick-start/README.md) section and follow
the instructions to set up a complete local environment. 
```

# Reference of plugin configurations on xwiki.cfg

Once you install the plugin in your XWiki instance, to make it work properly you will need to set up 

> a couple of properties.

Reviewer: ```suggestion
# Plugin configuration reference

Once you install the plugin in your XWiki instance, you will need to set up 
a couple of settings in your `xwiki.cfg` file to make it work properly.
```

## Required properties

> * `xwiki.authentication.authclass=com.xwiki.authentication.saml.XWikiSAML20Authenticator`

Reviewer: Add an explanation

* `xwiki.authentication.saml2.idp.single_sign_on_service.url`:

>     This option is used to tell what is the IDP URL where authentication happens. Usually, you can get this information from metadata.xml. Choose the "POST" URL.

Reviewer: ```suggestion
    This setting should point o the IDP authentication URL. Usually, you can get this information from a file named similarly to `metadata.xml`. Choose the "POST" URL.
```

* `xwiki.authentication.saml2.idp.entityid`:
    The IDP entity ID is a unique string that identifies the IDP and is used by the

> authenticator to validate the IDP. You usually get this information from a `metadata.xml` file.

Reviewer: ```suggestion
authenticator to validate the IDP. You usually get this information from the aforementioned `metadata.xml` file.
```

* `xwiki.authentication.saml2.sp.entityid`:
    SP stands for "Service Provider", in this case the XWiki instance. The SP entity ID
is a unique string (it can be any arbitrary value chosen by you) that identifies
your XWiki instance. Some people recommend using the application URL.

    Usually, you will configure this ID somewhere in the IDP, so it can tell whether authorization 
requests are coming from a valid application and to choose the correct configuration.
* `xwiki.authentication.saml2.idp.x509cert`
    This is the IDP certificate. This is used to validate the identity of the IDP provider

> to ensure it is legitimate.

Reviewer: This can be confusing. Is this the absolute or relative path to the certificate? Should I inline the certificate and set it here? Make this slightly more clear.

* `xwiki.authentication.saml2.sp.assertion_consumer_service.url`:

>     This is XWiki's URL that accepts the post to process the authentication coming from IDP. It will always be like this:`https://<your wiki domain>/bin/loginsubmit/XWiki/XWikiLogin`  

Reviewer: ```suggestion
    This is XWiki's URL that accepts the POST request to process the authentication coming from IDP. It will always be like this:`https://<your wiki domain>/bin/loginsubmit/XWiki/XWikiLogin`  
```

> ## Properties with default values

Reviewer: The properties above are a list and below are 3rd titles. Standardize.

The default value means:

+ IDP's `email` field will be mapped to Xwiki's `email`
+ IDP's `firstName` field will be mapped to Xwiki's `first_name`
+ IDP's `lastName` field will be mapped to Xwiki's `last_name`

The above fields are the minimum required fields.

### Default group for new users
`xwiki.authentication.saml2.defaultGroupForNewUsers=XWiki.SamlGroup`

You can set up the default group for users coming from the IDP. 

### The name of the attribute used to cache the authentication result in the current session
`xwiki.authentication.saml2.auth_field=saml_user`

You usually don't need to bother with this configuration. It is used to save a custom field
in the XWiki's users, mapping the IDP's user identity into it. Once you start with a value,
we don't recommend changing it. 

### List of fields to use for generating an XWiki username
`xwiki.authentication.saml2.xwiki_user_rule=first_name,last_name`

This property controls how the plugin will generate the username in XWiki, based on the user's
field values.

### Capitalize each field value when generating the username
`xwiki.authentication.saml2.xwiki_user_rule_capitalize=true`

> Controls whether username generation should capitalize each of the  xwiki_user_rule's attributes composition

Reviewer: ```suggestion
Controls whether username generation should capitalize each of the  xwiki_user_rule's attributes composition.
```

### NameIDFormat format
`xwiki.authentication.saml2.sp.nameidformat=urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress`

This field determines which IDP's field will be used to identify the user in XWiki. When a user authenticates, 

> this value will be stored in the field specified by `xwiki.authentication.saml2.auth_field` property

Reviewer: ```suggestion
this value will be stored in the field specified by `xwiki.authentication.saml2.auth_field` property.
```
